src/Aspire.Hosting/Exec/ExecResourceManager.cs

```diff
@@ -0,0 +1,188 @@
+    private readonly ResourceNotificationService _resourceNotificationService;
+
+    private string? _dcpExecResourceName;
+    private IResource? _execResource;
+
+    public ExecResourceManager(
+        ILogger<ExecResourceManager> logger,
+        IOptions<ExecOptions> execOptions,
+        DcpNameGenerator dcpNameGenerator,
+        DistributedApplicationModel model,
+        ResourceLoggerService resourceLoggerService,
+        ResourceNotificationService resourceNotificationService)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _execOptions = execOptions.Value;
+
+        _dcpNameGenerator = dcpNameGenerator ?? throw new ArgumentNullException(nameof(dcpNameGenerator));
+        _model = model ?? throw new ArgumentNullException(nameof(model));
+
+        _resourceLoggerService = resourceLoggerService ?? throw new ArgumentNullException(nameof(resourceLoggerService));
+        _resourceNotificationService = resourceNotificationService ?? throw new ArgumentNullException(nameof(resourceNotificationService));
+    }
+
+    public async IAsyncEnumerable<CommandOutput> StreamExecResourceLogs([EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        if (!_execOptions.Enabled)
+        {
+            yield break;
+        }
+        if (_execResource is null || string.IsNullOrEmpty(_dcpExecResourceName))
+        {
+            _logger.LogInformation("Exec resource can't be determined.");
+            yield break;
+        }
+
+        string type = "waiting";
+
+        // wait for the exec resource to be running
+        _ = Task.Run(async () =>
+        {
+            await _resourceNotificationService.WaitForResourceAsync(_execResource.Name, targetState: KnownResourceStates.Running, cancellationToken).ConfigureAwait(false);
+            type = "running";
+        }, cancellationToken);
+
+        // wait for the exec resource to reach terminal state
+        _ = Task.Run(async () =>
+        {
+            await _resourceNotificationService.WaitForResourceAsync(_execResource.Name, targetStates: KnownResourceStates.TerminalStates, cancellationToken).ConfigureAwait(false);
+            _resourceLoggerService.Complete(_dcpExecResourceName); // complete stops the `WatchAsync` async-foreach below
+        }, cancellationToken);
+        
+        await foreach (var logs in _resourceLoggerService.WatchAsync(_dcpExecResourceName).WithCancellation(cancellationToken).ConfigureAwait(false))
+        {
+            foreach (var log in logs)
+            {
+                yield return new CommandOutput
+                {
+                    Text = log.Content,
+                    IsErrorMessage = log.IsErrorMessage,
+                    LineNumber = log.LineNumber,
+                    Type = type
+                };
+            }
+        }
+    }
+
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
```

Comment:
I am not sure that we should be adding the resource in this background service. I think it would be better if we can hook an event somewhere earlier in the lifecycle for the target resource and add the exec resource - that way we might be able to avoid the changes that have been made to the DcpExecutor (no need to refresh the resource list).